### DISCUSSION TITLE
Mancato riconoscimento del body

### DISCUSSION THREAD

**Discussion by Clivio74** [fattureincloud/api#88] [Q&A] [answered]
Salve , sono uno sviluppatore junior in ambiente FileMaker, ho la necessità di risolvere un rognoso problema che espongo di seguito.
Inviando una chiamata POST oppure PUT contenente un JSON nella sezione body, ad esempio per la creazione di una fattura o la modifica della stessa, il sistema mi restituisce sempre l'errore:
{
	"error" : 
	{
		"message" : "Invalid request.",
		"validation_result" : 
		{
			"data" : [ "The i dati field is required." ]
		}
	}
}

L'applicazione da cui lancio la chiamata è FileMakerPro v19.5
Qualsiasi altra chiamata che non prevede un body funziona perfettamente.
Utilizzando Postman per eseguire delle verifiche , ho notato che ottengo il medesimo errore se elimino dall' header il parametro content-length.
Normalmente inviando dei dati JSON non ho mai avuto la necessità di inserire il parametro content-length, ho quindi provato ad inserirlo ma ottengo comunque lo stesso errore in risposta se invio la chiamata da FileMaker.
Chiedo quindi al team degli sviluppatori se il content-length è obbligatorio, se è possibile ignorarlo in qualche modo inserendo altre opzioni.
Il JSON che invio l'ho fatto validare su diverse piattaforme ed è corretto.
Allego di seguito il JSON inviato per una chiamata di tipo PUT, che restituisce l'errore indicato se inviata da FileMaker, invece viene eseguita correttamente se eseguita in Postman (che però calcola il content-length al momento dell'invio, quindi non ho modo di vederlo). La semplice modifica di una data
URL:
https://api-v2.fattureincloud.it/c/[il_codice_azienda]/issued_documents/[id_del_documento]

dati header:
-X POST 
-H "Content-Type: application/JSON" 
-H "Accept: application/JSON" 
-H "Authorization: Bearer [il token permanente]" 
-H "Content-Length: 30"

il JSON del body:
{
	"data" : 
	{
		"date" : "2022-08-18"
	}
}

**Comment by emajo** [comment]
Ciao @Clivio74, come standard HTTP il `content-length` è obbligatorio quando si fanno richieste con body, tuttavia nella maggior parte dei casi viene settato automaticamente dal client che effettua le chiamate, per quanto riguarda la PUT, non basta passare il dato modificato, ma è necessario passare tutta la risorsa. 

Consiglio di contattare il nostro supporto clienti così possiamo andare a controllare i log delle chiamate per vedere effettivamente dove risiede il problema.

Grazie,
Emanuele

**Reply by Clivio74** [reply]
Ho già contattato il supporto per questo problema e mi hanno risposto di scrivere qui.
Ma se per la PUT occorre tutta la risorsa, perchè se la eseguo allo stesso modo da POSTMAN funziona e mi modifica solo quel dato nella fattura ?
In ogni caso lo stesso errore mi ritorna anche con la POST di creazione della fattura (JSON verificato) e se la eseguo allo stesso modo da POSTMAN la fattura viene creata.
Riconosco che il fulcro della questione sta in FileMaker , ma da lì non ho la possibilità di vedere esattamente ciò che viene inviato al vostro sistema, in quanto eseguito da una funzione proprietaria alla quale io passo solo i parametri necessari. 
Con le API v1 non ho mai avuto nessun problema.

Grazie

**Reply by emajo** [reply]
Vero, errore mio la PUT funziona passando solo i dati da modificare, dai log riscontriamo che il body che arriva è vuoto, se riesce a condividere il codice con cui viene fatta la chiamata possiamo provare a indagare.

**Reply by Clivio74** [reply]
Giusta richiesta, vedo cosa posso fare, purtroppo non è cosa semplice in quanto FileMaker utilizza una funzione integrata alla quale vengono passati separatamente URL, header e body, ma sfortunatamente manca uno strumento per poter vedere cosa viene effettivamente inviato nello stream una volta composto il messaggio definitivo.
Mi consulto con i colleghi.

Grazie
Livio

**Comment by Luca-tr** [comment, marked as answer]
> -H "Content-Type: application/JSON" 

Ciao @Clivio74, ho avuto un problema simile e posso suggerirti, se ne hai la possibilità, di specificare il content-type in minuscolo nell'header della richiesta, cioè:
Content-Type: **application/json**

**Reply by Clivio74** [reply]
Ciao @Luca-tr 
Ho testato quanto da te suggerito ed il problema si è risolto. 
Ho diverse procedure che funzionano con il application/JSON in maiuscolo e quindi non ci avrei mai pensato.
Le GET ad esempio funzionano comunque anche su Fatture In Cloud

Grazie 1000 per il tuo aiuto , spero di poter ricambiare in futuro.

**Comment by DemetrioXY** [comment]
Ciao @Clivio74, posso approfittare per chiederti se hai un esempio di implementazione delle API 2.0 di FattureInCloud in FileMaker, o puoi dirmi dove posso trovarne e da dove partire nella documentazione per utilizzarle? Sono completamente a digiuno in questo campo.
Grazie!
